app/features/notes_generator/tools.py

```diff
@@ -0,0 +1,135 @@
+        for shape in slide.shapes:
+            if hasattr(shape, 'text'):
+                text.append(shape.text)
+    return clean_content("\n".join(text))
+
+def extract_text_from_md(file_path: str) -> str:
+    """Extract text from .md (Markdown) files."""
+    with open(file_path, 'r') as file:
+        return clean_content(file.read())
+
+def extract_text_from_xml(file_path: str) -> str:
+    """Extract text from .xml files."""
+    tree = ET.parse(file_path)
+    root = tree.getroot()
+    text = []
+    for elem in root.iter():
+        if elem.text:
+            text.append(elem.text.strip())
+    return clean_content("\n".join(text))
+
+def extract_text_from_xls(file_path: str) -> str:
+    """Extract text from .xls files."""
+    df = pd.read_excel(file_path, sheet_name=None)
+    text = []
+    for sheet in df.values():
+        text.append(sheet.to_string(index=False))
+    return clean_content("\n".join(text))
+
+def extract_text_from_xlsx(file_path: str) -> str:
+    """Extract text from .xlsx files."""
+    return extract_text_from_xls(file_path)  # Same as xls for now
+
+def extract_text_from_url(video_url: str) -> str:
+    """Extract text from a YouTube video URL."""
+    video_id = video_url.split("v=")[-1]
+    transcript = YouTubeTranscriptApi.get_transcript(video_id)
+    return clean_content(' '.join([item['text'] for item in transcript]))
+
+# Mapping file extensions to their handler functions
+file_handler_map = {
+    '.txt': extract_text_from_txt,
+    '.pdf': extract_text_from_pdf,
+    '.docx': extract_text_from_docx,
+    '.csv': extract_text_from_csv,
+    '.pptx': extract_text_from_pptx,
+    '.md': extract_text_from_md,
+    '.xml': extract_text_from_xml,
+    '.xls': extract_text_from_xls,
+    '.xlsx': extract_text_from_xlsx
+}
+
+def extract_text_from_file(file_path: str) -> str:
+    """Extract text from a file based on its extension using the mapping."""
+    file_extension = os.path.splitext(file_path)[1].lower()
+
+    # Check if the file extension is in our handler map
+    if file_extension in file_handler_map:
+        handler = file_handler_map[file_extension]()  # Instantiate the handler
+        return handler(file_path)
+    else:
+        raise ValueError(f"Unsupported file type: {file_extension}")
+
+# Function to generate notes with RAGB from the extracted content
+def generate_notes_rag(content: str, focus_topic: Optional[str], structure: str) -> str:
+    # 1. Create embeddings and vector store from content (one-time or repeated)
+    embeddings = OpenAIEmbeddings(openai_api_key="OPEN_AI_API_KEY")
+    # This can handle large content by splitting it into documents
+    docs = [content]  # If you split content into chunks, store them as multiple docs
+    vectorstore = FAISS.from_texts(docs, embeddings)
+
+    # 2. Create a QA chain using your vectorstore
+    retriever = vectorstore.as_retriever(search_kwargs={"k": 3})
+    qa_chain = RetrievalQA.from_chain_type(
```

Comment:
Why is `RetrievalQA` needed? This class is deprecated since LangChain 0.1.17: https://python.langchain.com/api_reference/langchain/chains/langchain.chains.retrieval_qa.base.RetrievalQA.html

Reply:
Thank you